tm2/pkg/std/memfile.go

```diff
@@ -2,99 +2,231 @@ package std
 
 import (
 	"fmt"
+	"io/ioutil"
+	"path/filepath"
 	"regexp"
+	"slices"
 	"sort"
 	"strings"
 )
 
+// XXX rename to mempackage.go
+
+const fileNameLimit = 256
+const pkgNameLimit = 256
+const pkgPathLimit = 256
+
+var (
+	// See also gnovm/pkg/gnolang/validate_mempackage.go.
+	reFileName   = regexp.MustCompile(`^(([a-z0-9_]+|[A-Z0-9_]+)\.[a-z0-9_]{1,7}|LICENSE|README)$`)
+	rePkgName    = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)
+	rePkgPathURL = regexp.MustCompile(`^([a-z0-9-]+\.)*[a-z0-9-]+\.[a-z]{2,}\/(?:p|r)(?:\/_?[a-z]+[a-z0-9_]*)+$`)
+	rePkgPathStd = regexp.MustCompile(`^([a-z][a-z0-9_]*\/)*[a-z][a-z0-9_]+$`)
+)
+
+//----------------------------------------
+// MemPackage
+
+// A MemFile is the simplest representation of a "file".
+//
+// File Name must contain a single dot and extension.
+// File Name may be ALLCAPS.xxx or lowercase.xxx; extensions lowercase.
+// e.g. OK:     README.md, readme.md, readme.txt, READ.me
+// e.g. NOT OK: Readme.md, readme.MD, README, .readme
+// File Body can be any string.
+//
+// NOTE: It doesn't have owners or timestamps. Keep this as is for portability.
+// Not even date created, ownership, or other attributes.  Just a name, and a
+// body.  This keeps things portable, easy to hash (otherwise need to manage
+// e.g. the encoding and portability of timestamps).
 type MemFile struct {
 	Name string `json:"name" yaml:"name"`
 	Body string `json:"body" yaml:"body"`
 }
 
+func (mfile *MemFile) ValidateBasic() error {
+	if len(mfile.Name) > fileNameLimit {
+		return fmt.Errorf("name length %d exceeds limit %d", len(mfile.Name), fileNameLimit)
+	}
+	if !reFileName.MatchString(mfile.Name) {
+		return fmt.Errorf("invalid file name %q", mfile.Name)
+	}
+	return nil
+}
+
+// Print file to stdout.
+func (mfile *MemFile) Print() error {
+	fmt.Printf("MemFile[%q]:\n", mfile.Name)
+	fmt.Println(mfile.Body)
+	return nil
+}
+
+//----------------------------------------
+// MemPackage
+
 // MemPackage represents the information and files of a package which will be
 // stored in memory. It will generally be initialized by package gnolang's
 // ReadMemPackage.
```

Comment:
```suggestion
// MemPackage represents the information and files of a package which will be
// stored in memory. It will generally be initialized by package gnolang's
// ReadMemPackage.
// Note: a package does not support subfolders.
```